### PR TITLE
Converts the APC console from html to TGUI

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

I've converted the apc control console into tgui, and made it much easier to use. 
Yes this is a ce buff, we should tune the behavior of the tool, not it's ease of use.

I've up'd the access requirements to CE and changed how the emag effects it somewhat, it removes logging entirely rather then just disabling the name.

I took some notes/code from the power monitor console, I don't intend to rip it off wholesale, but I was/am bad at this and needed a base. Thank you @actioninja/stylemistake. 

RE: performance, it's a little less then twice as slow as the power monitor console to load, I assume that's due to all the buttons.
image
image
image

## Why It's Good For The Game

This is a useful tool, it should be actually usable.

## Changelog
:cl:
tweak: The APC control console is now tgui rather then html. It should be a lot more usable. I've made other changes, you won't notice.
/:cl:

### CODE REVIEW

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+const LogPanel = (props, context) => {
+  const { data } = useBackend(context);
+
+  const logs = flow([
+    map((line, i) => ({
+      ...line,
+      // Generate a unique id
+      id: line.entry + i,
+    })),
+  ])(data.logs.reverse());
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/ApcControl.js:259]: Just saying that with `flow`, you have an option to structure operations on data as a series of transforms, which is quite nice and makes it easier to reorder or add new transformations when necessary.

Another thing to watch out, is that `Array.prototype.reverse()` is a mutating method, which means, that `data.logs` will become reversed, and not just the return value. By reversing the new object that we return with `map`, you avoid mutating `data.logs`.

```suggestion
  const logs = flow([
    map((line, i) => ({
      ...line,
      // Generate a unique id
      id: line.entry + i,
    })),
    logs => logs.reverse(),
  ])(data.logs);
```

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+              onClick={() => act('breaker',
+                { ref: apc.ref }
+              )}
+            />
+          </Table.Cell>
+          <Table.Cell>
+            <Button
+              onClick={() => act('access-apc',
+                { ref: apc.ref }
+              )}>
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/ApcControl.js:207]: Write it this way please.

```suggestion
              onClick={() => act('access-apc', {
                ref: apc.ref,
              })}>
```

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+        <Table.Row header
+          key={apc.id}
+          className="Table__row candystripe">
+          <Table.Cell>
+            <Button
+              icon={apc.operating ? 'power-off' : 'times'}
+              color={apc.operating ? 'good' : 'bad'}
+              onClick={() => act('breaker',
+                { ref: apc.ref }
+              )}
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/ApcControl.js:200]:
```suggestion
              onClick={() => act('breaker', {
                ref: apc.ref,
              })}
```

**Review by stylemistake — DISMISSED** [review]
Overall, there is a lot of naming inconsistency between your CSS and JS, and some of them are poorly named. Some of the buttons use `fixed` positioning, which is very concerning, because parent element is positioned relatively to the document, and probably breaks the bounds of the parent element; `absolute` would be preferred, and also coordinates in `px` as well.

You need to do quite a lot of homework on this one.

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+          onClick={() => {
+            setTabIndex(2);
+            act("check-logs");
+          }}>
+          Log View Panel
+        </Tabs.Tab>
+      </Tabs>
+      {data.restoring === 1&& (
+        <Dimmer fontSize="32px">
+          <Icon name="cog" spin={1} />
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/ApcControl.js:64]: Spin is a boolean property.

1. Don't use numbers for booleans, this is not BYOND code.
2. Property without an assigned value is considered `true`, like in normal HTML code.

```suggestion
          <Icon name="cog" spin />
```

File: tgui/packages/tgui/styles/layouts/PowerHeader.scss [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+  margin-top: 98px;
+  font-family: Consolas, monospace;
+  font-size: 14px;
+}
+
+.LogWindowTab .Layout__content {
+  margin-top: 20px;
+  font-family: Consolas, monospace;
+  font-size: 14px;
+}
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/styles/layouts/PowerHeader.scss:11]: That is really useful, thank you

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+      {logs.map((line, i) => (
+        <Table.Row header
+          key={line.id}
+          className="Table__row candystripe">
+          <Table.Cell>
+            {line.entry}
+          </Table.Cell>
+        </Table.Row>
+      ))}
+    </Table>
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:271]: Aight gottcha. I've changed to Box, not gonna use the margin change cause it appears to be causing bloat on every second line.

I'm using inline and minWidth="100%" to get the box to fully fill the screen. If there's a better way of doing that please let me know. I thought about using Flex but I thought that might be too laggy.

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+              : 'red'
+          )
+          || charging === 1 && 'yellow'
+          || charging === 2 && 'green'
+        )} />
+      <Box
+        inline
+        width="36px"
+        textAlign="right">
+        { toFixed(charge) + '%'}
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:303]: gottcha

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+        <Table.Cell collapsing title="Equipment">
+          Eqp
+        </Table.Cell>
+        <Table.Cell collapsing title="Lighting">
+          Lgt
+        </Table.Cell>
+        <Table.Cell collapsing title="Environment">
+          Env
+        </Table.Cell>
+      </Table.Row>
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:164]: You're right. Is there a good way to get it to be static with the scrolling table? In specific environment lighting and equipment can be hard to remember the order of, and I'd like that info to be on hand.

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+    sortByField === 'draw' && sortBy(
+      apc => -powerRank(apc.load),
+      apc => -parseFloat(apc.load)),
+  ])(data.apcs);
+  return (
+    <Table>
+      {apcs.map((apc, i) => (
+        <Table.Row header
+          key={apc.id}
+          className="Table__row candystripe">
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:193]: I see. I converted it to <Table> in a vain attempt to fix the stuff you spotted below.

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+        </div>
+      )}
+    </Fragment>
+  );
+};
+
+const powerRank = str => {
+  const unit = String(str.split(' ')[1]).toLowerCase();
+  return ['w', 'kw', 'mw', 'gw'].indexOf(unit);
+};
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:92]: That and the AreaCharge, got it

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+          onClick={() => setSortByField(
+            sortByField !== 'charge' && 'charge'
+          )} />
+        <Button.Checkbox
+          checked={sortByField === 'draw'}
+          content="Draw"
+          onClick={() => setSortByField(sortByField !== 'draw' && 'draw')} />
+        {emagged === 1 && (
+          <Fragment>
+            <Button color={(logging === 1 ? "bad" : "good")}
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:124]: Understood

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+
+const powerRank = str => {
+  const unit = String(str.split(' ')[1]).toLowerCase();
+  return ['w', 'kw', 'mw', 'gw'].indexOf(unit);
+};
+
+const ControlPanel = (props, context) => {
+  const { act, data } = useBackend(context);
+  const emagged = data.emagged;
+  const logging = data.logging;
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:97]: alright

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 13]
```diff
@@ -0,0 +1,335 @@
...
+        </Dimmer>
+      )}
+      {tabIndex === 1 && (
+        <Fragment>
+          <ControlPanel />
+          <div className="PowerWindowHeader">
+            <Window.Content scrollable>
+              <ApcControlScene />
+            </Window.Content>
+          </div>
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:75]: I'm using `fillPositionedParent` and manually setting top, is that alright? It appears to require a surrounding <Box>

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+        </Fragment>
+      )}
+      {tabIndex === 2 && (
+        <div className="LogWindowTab">
+          <Window.Content scrollable>
+            <LogPanel />
+          </Window.Content>
+        </div>
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:83]: See above

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 15]
```diff
@@ -0,0 +1,335 @@
+import { Box, Section, Button, Table, Icon, Dimmer, Tabs } from '../components';
+import { map, sortBy } from 'common/collections';
+import { flow } from 'common/fp';
+import { toFixed } from 'common/math';
+import { Window, Fragment } from '../layouts';
+import { useBackend, useLocalState } from '../backend';
+import { pureComponentHooks } from 'common/react';
+import { FlexItem } from '../components/Flex';
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:8]: Handy

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+          }}>
+          Log View Panel
+        </Tabs.Tab>
+      </Tabs>
+      {data.restoring === 1&& (
+        <Dimmer fontSize="32px">
+          <Icon name="cog" spin={1} />
+          {' Resetting...'}
+        </Dimmer>
+      )}
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:67]: Will do

[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:67]: Oh and the dimmer looks about like this currently 
image

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 17]
```diff
@@ -0,0 +1,335 @@
...
+          checked={sortByField === 'draw'}
+          content="Draw"
+          onClick={() => setSortByField(sortByField !== 'draw' && 'draw')} />
+        {emagged === 1 && (
+          <Fragment>
+            <Button color={(logging === 1 ? "bad" : "good")}
+              content={(logging === 1 ? "Stop Logging" : "Restore Logging")}
+              onClick={() => act('toggle-logs')}
+              right={13}
+              position="fixed"
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:128]: I've cleared it out

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 18]
```diff
@@ -0,0 +1,335 @@
...
+            />
+            <Button content="Reset Console"
+              onClick={() => act('restore-console')}
+              right={5}
+              position="fixed"
+            />
+          </Fragment>
+        )}
+        <Button color={'bad'}
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:137]: Alright. Is that a style choice? Or does '' vs "" matter

File: tgui/packages/tgui/interfaces/ApcControl.js (hunk 1 above)
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:259]: That is helpful, thanks

File: tgui/packages/tgui/interfaces/ApcControl.js (hunk 2 above)
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:207]: Alright

File: tgui/packages/tgui/interfaces/ApcControl.js (hunk 3 above)
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:200]: Gottcha

File: tgui/packages/tgui/interfaces/ApcControl.js (hunk 4 above)
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:64]: Thank you

File: tgui/packages/tgui/interfaces/ApcControl.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:193]: This compares with the monitor console at 24 for initial, 38 for full load

[the remark this answers is not in the document]
Comment (stylemistake) [reply tgui/packages/tgui/interfaces/ApcControl.js]: Yeah, that's chonky, but oh well.

File: tgui/packages/tgui/interfaces/ApcControl.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (stylemistake) [reply tgui/packages/tgui/interfaces/ApcControl.js]: There is no good way to keep it static. There are certain css rules for this, but I would just keep using what's available out of the box until this is standardized.

File: tgui/packages/tgui/interfaces/ApcControl.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (stylemistake) [reply tgui/packages/tgui/interfaces/ApcControl.js]: That works very nicely!

File: tgui/packages/tgui/interfaces/ApcControl.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (stylemistake) [reply tgui/packages/tgui/interfaces/ApcControl.js]: Semantically `prop="string"` means it's a constant string property. It doesn't really affect the way it compiles (`prop="string"` and `prop={'string'}` will result in same output), so yeah, it's pretty much a style/semantics choice.

File: tgui/packages/tgui/interfaces/ApcControl.js [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+        </Table.Cell>
+        <Table.Cell collapsing title="Lighting">
+          Lgt
+        </Table.Cell>
+        <Table.Cell collapsing title="Environment">
+          Env
+        </Table.Cell>
+      </Table.Row>
+      {apcs.map((apc, i) => (
+        <tr header
```
Comment (stylemistake) [inline tgui/packages/tgui/interfaces/ApcControl.js:191]: remove header prop, also missing Table__row class (although it won't do much, i think).

Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js:191]: Gottcha

**Comment by LemonInTheDark** [comment]
Still can't get over losing the fixed headers, but you're right, they looked like shit 
image

File: tgui/packages/tgui/interfaces/ApcControl.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js]: aight. doesn't matter overmuch, but that's a shame

File: tgui/packages/tgui/interfaces/ApcControl.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply tgui/packages/tgui/interfaces/ApcControl.js]: that's good to know, thanks

**Comment by stylemistake** [comment]
Try using position="sticky". If that works, would be nice to add a "sticky" prop to Table.Row, which sets a Table__row--sticky class, with approproate CSS rules.

https://stackoverflow.com/a/52008347

**Comment by LemonInTheDark** [comment]
Not sure if my process is wrong or something, but I can't get it to work. Might be how Window.Content scrollable works or something

**Review by stylemistake — APPROVED** [review]

**Comment by kriskog** [comment]
merge conflict

**Pull request merged**